crowbar_framework/app/models/service_object.rb

```diff
@@ -1113,7 +1144,17 @@ def apply_role(role, inst, in_queue)
         # If new_nodes is empty, we are just removing the proposal.
         unless new_nodes.empty?
           new_nodes.each do |node_name|
-            node = NodeObject.find_node_by_name(node_name)
+            # skip adding nodes to the batch unless they are really in the list to be deployed
+            # do it before the Node load to avoid doing the call if the node is not there,
+            # as we dont want to spend cycles doing extra calls not needed
+            # This also means we can't ensure that the node has all required roles
+            # through the use of pending_node_actions
+            # It's a reasonable trade-off in the context of this specific optimization,
+            # as the nodes should already have all roles, unless the customer removes roles
+            # manually or the roles disappear from the node magically (bugs)
+            next unless new_elements[role_name].include?(node_name)
+
+            pre_cached_nodes[node_name] ||= NodeObject.find_node_by_name(node_name)
```

Comment:
we are missing a node assignment here:

```
node = pre_cached_nodes[node_name] ||= NodeObject.find_node_by_name(node_name)
```

because we check `node` in line 1170

Reply:
or you could change line 1170 to be: `if pre_cached_nodes[node_name].nil?` which is how it is on 4.0/master